library/eddsa.c

```diff
@@ -0,0 +1,446 @@
+
+    if( ctx == NULL || hash == NULL || sig == NULL || slen == NULL || f_rng == NULL )
+        return( MBEDTLS_ERR_ECP_BAD_INPUT_DATA );
+
+    mbedtls_mpi_init( &r );
+    mbedtls_mpi_init( &s );
+
+    MBEDTLS_MPI_CHK( mbedtls_eddsa_sign( &ctx->grp, &r, &s, &ctx->d,
+                                                 hash, hlen, f_rng,
+                                                 p_rng, eddsa_id, ed_ctx,
+                                                 ed_ctx_len ) );
+
+    MBEDTLS_MPI_CHK( eddsa_signature_to_asn1( &r, &s, sig, sig_size, slen ) );
+
+cleanup:
+    mbedtls_mpi_free( &r );
+    mbedtls_mpi_free( &s );
+
+    return( ret );
+#endif /* MBEDTLS_ASN1_WRITE_C */
+}
+
+/*
+ * Restartable read and check signature
+ */
+int mbedtls_eddsa_read_signature( mbedtls_ecp_keypair *ctx,
+                          const unsigned char *hash, size_t hlen,
+                          const unsigned char *sig, size_t slen,
+                          int (*f_rng)(void *, unsigned char *, size_t),
+                          void *p_rng, mbedtls_eddsa_id eddsa_id,
+                          const unsigned char *ed_ctx, size_t ed_ctx_len )
+{
+    int ret = MBEDTLS_ERR_ERROR_CORRUPTION_DETECTED;
+    unsigned char *p = (unsigned char *) sig;
+    const unsigned char *end = sig + slen;
+    size_t len;
+    mbedtls_mpi r, s;
+
+#if !defined(MBEDTLS_ASN1_PARSE_C)
+    return( MBEDTLS_ERR_ECP_FEATURE_UNAVAILABLE );
+#else
+
+    if( ctx == NULL || hash == NULL || sig == NULL )
+        return( MBEDTLS_ERR_ECP_BAD_INPUT_DATA );
+
+    mbedtls_mpi_init( &r );
+    mbedtls_mpi_init( &s );
+
+    if( ( ret = mbedtls_asn1_get_tag( &p, end, &len,
+                    MBEDTLS_ASN1_CONSTRUCTED | MBEDTLS_ASN1_SEQUENCE ) ) != 0 )
+    {
+        ret += MBEDTLS_ERR_ECP_BAD_INPUT_DATA;
+        goto cleanup;
+    }
+
+    if( p + len != end )
+    {
+        ret = MBEDTLS_ERROR_ADD( MBEDTLS_ERR_ECP_BAD_INPUT_DATA,
+              MBEDTLS_ERR_ASN1_LENGTH_MISMATCH );
+        goto cleanup;
+    }
+
+    if( ( ret = mbedtls_asn1_get_mpi( &p, end, &r ) ) != 0 ||
+        ( ret = mbedtls_asn1_get_mpi( &p, end, &s ) ) != 0 )
+    {
+        ret += MBEDTLS_ERR_ECP_BAD_INPUT_DATA;
+        goto cleanup;
+    }
+
+    if( ( ret = mbedtls_eddsa_verify( &ctx->grp, hash, hlen,
+                                      &ctx->Q, &r, &s, f_rng, p_rng,
+                                      eddsa_id, ed_ctx, ed_ctx_len ) ) != 0 )
+        goto cleanup;
+
+    /* At this point we know that the buffer starts with a valid signature.
+     * Return 0 if the buffer just contains the signature, and a specific
+     * error code if the valid signature is followed by more data. */
+    if( p != end )
+        ret = MBEDTLS_ERR_ECP_SIG_LEN_MISMATCH;
+
+cleanup:
+    mbedtls_mpi_free( &r );
+    mbedtls_mpi_free( &s );
+
+    return( ret );
+#endif /* MBEDTLS_ASN1_PARSE_C */
+}
+
+/*
+ * Generate key pair
+ */
+int mbedtls_eddsa_genkey( mbedtls_ecp_keypair *ctx, mbedtls_ecp_group_id gid,
```

Comment:
I am not sure about the usefulness of this function. It is basically the same as `mbedtls_ecp_gen_key` with argument swapped.

Reply:
I am not sure, too. My initial approach was to mimmic ecdsa.c with EdDSA. Probably it will be removed.

Reply:
Yes, I understand from where it comes from. The difference is that in the ECDSA case, the first argument is actually a pointer to `mbedtls_ecdsa_context` instead of a pointer to a `mbedtls_ecp_keypair`.